### DISCUSSION TITLE
Chat Stream Community Plugin AI Threads

### DISCUSSION THREAD

**Discussion by VentureTactics** [rpggio/obsidian-chat-stream#24] [Show and tell]
## Summary
In the following image I've used Chat Stream to link relevant plugin dev docs as explicit context for it's answers:
![Chat Stream AI Prompt - Obsidian Queries 1](https://github.com/rpggio/obsidian-chat-stream/assets/88736327/5bbce614-0c13-4131-b22b-dec1e41dd91b)

## Reason
It's easy to forget the exact structure needed for each plugin (especially complex ones like Dataview, Templater, and similar).
When I simply vectorize my entire vault, it misses critical details that can only be accomplished by having the entire files in the actual context window, so I needed a way to better target the information for proper outputs.

## Workflow

1. I created MD files of the dev docs and git README of almost every plugin I'm currently using (I did this manually since I don't have a good way to automatically scrape every page in dev docs that aren't rendered in a single page)
2. I used the File-Index plugin to create a JSON outline of my file folder structure to help with Dataview queries and so forth.
3. I used Chat Stream to render a DV query of my plugins, from which I can use the Hover Editor to quickly drag new context files into my canvas.
4. I link the relevant information to my queries, and generate Assistant outputs as needed.
5. I can duplicate the Canvas as needed to apply a similar reference structure to other instances where I might want to cherry-pick files from a DV query to use as context in my AI outputs.

**Comment by rpggio** [comment]
Very interesting! You are pushing these tools to their limits. I wonder if a bridge between canvas and LangChain or promptchainer.io (for ex) would help people who want to go further.

Have you checked out [Cannoli?](https://github.com/DeabLabs/cannoli) Kind of a canvas meta-programming plugin.

**Reply by VentureTactics** [reply]
Oh, this is just the beginning, haha.
We're just scratching the surface. 
I think we can potentially unlock some really cool possibilities with this.
I've looked at Cannoli, but haven't tried it yet.
It seems like Cannoli can run a canvas like a workflow, however, Canvas Stream seems more equipped to handle train-of-thought style prompting.
Nodes can easily be re-arranged and re-connected, quickly changing the context window and custom selecting which node to run the generation on.
That's not to say Canvas Stream can't handle pre-defined workflows though, with a bit of tailoring.
